cpp/src/Ice/ConnectionI.cpp

```diff
@@ -1476,25 +1430,35 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
 
 // executeFromThisThread dispatches to the correct DispatchQueue
 #ifdef ICE_SWIFT
-    _threadPool->executeFromThisThread(make_shared<ExecuteUpcall>(
-        shared_from_this(),
-        std::move(connectionStartCompleted),
-        sentCBs,
-        std::move(messageUpcall),
-        current.stream));
+    auto stream = make_shared<InputStream>();
+    current.stream.swap(*stream);
```

Comment:
Should we add a move constructor to InputStream?

It's probably reasonably easy since the base class (IceInternal::Buffer) has one. OutputStream already has a move constructor.

Reply:
I tried here but it didn't work because the lambda can't capture the non-const moved input stream (if I recall correctly!).

Reply:
I don't understand your comment.

The base class (IceInternal::Buffer) is not copy-constructible or copy-assignable:
```cpp
Buffer(const Buffer&) = delete;
Buffer& operator=(const Buffer&) = delete;
```

Same for the derived classes.

If you really need to capture the InputStream in a lambda, you can use a shared_ptr like you did in this PR (not great but works). If doable, it's preferable to not capture it by pass it around as a move-only parameter like I did for OutputStream/OutgoingResponse, e.g.
```cpp
virtual void dispatch(IncomingRequest& request, std::function<void(OutgoingResponse)> sendResponse);
```

In any case, having InputStream movable would just provide an alternative to swap:
```cpp
// No need to swap anything; messageStream is "reinitialized" after the move
auto stream = make_shared_ptr<InputStream>(std::move(messageStream));
```

Reply:
I'll check this next.